src/data/set/intervals/monotone.lean

```diff
@@ -22,34 +28,163 @@ section
   (h₂ : strict_anti_on f (Ici a)) : strict_anti f :=
 (h₁.dual_right.Iic_union_Ici h₂.dual_right).dual_right
+
+/-- If `f` is monotone both on `s` and `t`, with `s` to the left of `t` and the center
+point belonging to both `s` and `t`, then `f` is strictly monotone on `s ∪ t` -/
+protected lemma monotone_on.union' {s t : set α} {c : α} (h₁ : monotone_on f s)
+  (h₂ : monotone_on f t) (hs : s ⊆ Iic c) (ht : t ⊆ Ici c) (cs : c ∈ s) (ct : c ∈ t) :
+  monotone_on f (s ∪ t) :=
+begin
+  have A : ∀ x, x ∈ s ∪ t → x ≤ c → x ∈ s,
+  { assume x hx hxc,
+    cases hx, { exact hx },
+    rcases eq_or_lt_of_le hxc with rfl|h'x, { exact cs },
+    exact (lt_irrefl _ (h'x.trans_le (ht hx))).elim },
+  have B : ∀ x, x ∈ s ∪ t → c ≤ x → x ∈ t,
+  { assume x hx hxc,
+    cases hx, swap, { exact hx },
+    rcases eq_or_lt_of_le hxc with rfl|h'x, { exact ct },
+    exact (lt_irrefl _ (h'x.trans_le (hs hx))).elim },
+  assume x hx y hy hxy,
+  rcases lt_or_le x c with hxc|hcx,
+  { have xs : x ∈ s, from A _ hx hxc.le,
+    rcases lt_or_le y c with hyc|hcy,
+    { exact h₁ xs (A _ hy hyc.le) hxy },
+    { exact (h₁ xs cs hxc.le).trans (h₂ ct (B _ hy hcy) hcy) } },
+  { have xt : x ∈ t, from B _ hx hcx,
+    have yt : y ∈ t, from B _ hy (hcx.trans hxy),
+    exact h₂ xt yt hxy }
+end
+
+/-- If `f` is monotone both on `(-∞, a]` and `[a, ∞)`, then it is monotone on the whole line. -/
 protected lemma monotone_on.Iic_union_Ici (h₁ : monotone_on f (Iic a))
   (h₂ : monotone_on f (Ici a)) : monotone f :=
 begin
-  intros x y hxy,
-  cases le_total x a with hxa hax; [cases le_total y a with hya hay, skip],
-  exacts [h₁ hxa hya hxy, (h₁ hxa le_rfl hxa).trans (h₂ le_rfl hay hay), h₂ hax (hax.trans hxy) hxy]
+  rw [← monotone_on_univ, ← @Iic_union_Ici _ _ a],
+  exact monotone_on.union' h₁ h₂ subset.rfl subset.rfl le_rfl le_rfl,
 end
 
+/-- If `f` is strictly antitone both on `s` and `t`, with `s` to the left of `t` and the center
+point belonging to both `s` and `t`, then `f` is strictly antitone on `s ∪ t` -/
+protected lemma antitone_on.union' {s t : set α} {c : α} (h₁ : antitone_on f s)
+  (h₂ : antitone_on f t) (hs : s ⊆ Iic c) (ht : t ⊆ Ici c) (cs : c ∈ s) (ct : c ∈ t) :
+  antitone_on f (s ∪ t) :=
+(h₁.dual_right.union' h₂.dual_right hs ht cs ct).dual_right
+
+/-- If `f` is antitone both on `(-∞, a]` and `[a, ∞)`, then it is antitone on the whole line. -/
 protected lemma antitone_on.Iic_union_Ici (h₁ : antitone_on f (Iic a))
   (h₂ : antitone_on f (Ici a)) : antitone f :=
 (h₁.dual_right.Iic_union_Ici h₂.dual_right).dual_right
 
+/-- If a function is monotone on a set `s`, then it admits a monotone extension to the whole space
+provided `s` has a smallest element `a` and a largest element `b`. -/
+lemma monotone_on.exists_monotone_extension {β : Type*} [conditionally_complete_linear_order β]
+  {f : α → β} {s : set α} (h : monotone_on f s) {a b : α}
+  (as : a ∈ s) (bs : b ∈ s) (hab : s ⊆ Icc a b) :
```

Comment:
This might be more conveniently formulated with `is_least` and `is_greatest`? But feel free to leave it as-is.